### ISSUE TITLE
badarith,[{erlang,'+',[infinity,64000]

### ISSUE THREAD

**Issue by stofel** [ninenines/cowboy#1411]
On load testing current master tag 3000rps get crashes
=ERROR REPORT==== 30-Oct-2019::09:55:08.485504 ===
Ranch listener my_http_listener had connection process started with cowboy_clear:start_link/4 at <0.29776.42> exit with reason: {badarith,[{erlang,'+',[infinity,64000],[]},{cowboy_http,commands,3,[{file,"/home/alexs/work/psp/_build/default/lib/cowboy/src/cowboy_http.erl"},{line,948}]},{cowboy_http,loop,1,[{file,"/home/alexs/work/psp/_build/default/lib/cowboy/src/cowboy_http.erl"},{line,234}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}

**Comment by essen** [comment]
It's missing a clause for when a flow (read body) command arrives after the body was finished reading.

**Comment by stofel** [comment]
And?
I use
`
init(Req, Opts) ->
 M = cowboy_req:method(Req),
 B = cowboy_req:read_urlencoded_body(Req)
 ...
`
read_urlencoded_body crashed
Is here need some patch?

**Comment by essen** [comment]
I'm just saying why there's a crash, I'll work on a fix later.

**Comment by roivinir** [comment]
Seeing the same crash as @stofel also under heavy load 
Also a very similar one soon there after. Looks like the same issue.

"Ranch listener ApiServicesWeb.Endpoint.HTTP had connection process started with :cowboy_clear:start_link/4 at #PID<0.19133.8> exit with reason: {:badarith, [{:erlang, :-, [:infinity, 1460], []}, {:cowboy_http, :update_flow, 3, [file: '/home/.../deps/cowboy/src/cowboy_http.erl', line: 381]}, {:cowboy_http, :after_parse, 1, [file: '/home/.../deps/cowboy/src/cowboy_http.erl', line: 358]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}]}\n",

**Comment by essen** [comment]
Feel free to experiment with a fix I don't think I'll get to it this week.

**Comment by essen** [comment]
Please try this patch:

``` patch
diff --git a/src/cowboy_http.erl b/src/cowboy_http.erl
index f128a44..d60f4e0 100644
--- a/src/cowboy_http.erl
+++ b/src/cowboy_http.erl
@@ -938,6 +938,9 @@ commands(State=#state{out_streamid=Current, streams=Streams0}, StreamID, Command
    Streams = lists:keyreplace(StreamID, #stream.id, Streams0,
        Stream#stream{queue=Queue ++ Commands}),
    State#state{streams=Streams};
+%% When we have finished reading the request body, do nothing.
+commands(State=#state{flow=infinity}, StreamID, [{flow, _}|Tail]) ->
+   commands(State, StreamID, Tail);
 %% Read the request body.
 commands(State=#state{socket=Socket, transport=Transport, flow=Flow0}, StreamID,
        [{flow, Size}|Tail]) ->

**Comment by roivinir** [comment]
Tried this patch. Seems to have fixed half of the problem. Still see the issue on line 381 

"message": "Ranch listener ApiServicesWeb.Endpoint.HTTP had connection process started with :cowboy_clear:start_link/4 at #PID<0.20703.5> exit with reason: {:badarith, [{:erlang, :-, [:infinity, 1460], []}, {:cowboy_http, :update_flow, 3, [file: '/home/travis/build/ni-kismet/helium-dataservices/deps/cowboy/src/cowboy_http.erl', line: 381]}, {:cowboy_http, :after_parse, 1, [file: '/home/travis/build/ni-kismet/helium-dataservices/deps/cowboy/src/cowboy_http.erl', line: 358]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}]}\n"

**Comment by essen** [comment]
That's not supposed to happen, if it occurs it must be because of a different bug. The value is set to `infinity` when the stream was finished reading or its body will be skipped (so the stream is removed from the list). But this error occurs in code where the stream was not removed?

Do you have any other errors or some more information about the request that's causing this?

**Comment by alexandre-kivra** [comment]
We are also seeing from occasionally the `badarith` error on the call to `update_flow/3`.

We don't have much context on the error. What information would you need to investigate further? Maybe I can patch cowboy to add some more logging and let it run for a while until the error happens again.

**Comment by essen** [comment]
The state would be great. Maybe just adding `when Flow0 =/= infinity` would be enough to have the state in the error report, otherwise logging when `Flow0 =:= infinity`.

**Comment by huss-kivra** [comment]
We added the guard and got the crash:
```
{function_clause,
 [{cowboy_http,update_flow,
   [nofin,<<"DATA">>,
    {state,<0.2419.0>,core_http,#Port<0.63941>,ranch_tcp,undefined,
     #{connection_type => supervisor,
       env => #{dispatch => [...]},
       idle_timeout => 300000,
       max_keepalive => infinity,
       middlewares => [cowboy_router,cors_middleware,cowboy_handler],
       request_timeout => 300000},
     <<>>,
     #{},
     {{127,0,0,1},60276},
     {{127,0,0,1},8443},
     undefined,
     #Ref<0.3556285883.1641021443.193014>,
       233,
     {ps_body,195449,19940,fun cow_http_te:stream_identity/2,{19940,195449}},
     infinity,
     233,
     wait,
     infinity,
     [{stream,233,
       {cowboy_stream_h,
        {state,
         undefined,
         core_http,
         <0.32112.35>,
           undefined,
         undefined,
         undefined,
         undefined,
         0,
         nofin,
         <<"read_body_buffer, ending with DATA">>,
         19940,
         undefined,
         normal}},
       <<"POST">>,
       'HTTP/1.1',
       undefined,
       undefined,
       0,
       []}],
     [{child,<0.32112.35>,233,5000,undefined}]}],
   [{file,"/build/kivra_core/_build/default/lib/cowboy/src/cowboy_http.erl"},{line,378}]},{cowboy_http,after_parse,1,[{file,"/build/kivra_core/_build/default/lib/cowboy/src/cowboy_http.erl"},{line,358}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
```

**Comment by essen** [comment]
It probably happens when two streams are running concurrently, I'll try to reproduce.

**Comment by essen** [comment]
That's it, I can reproduce.

**Comment by essen** [comment]
Please try the following patch:

``` patch
diff --git a/src/cowboy_http.erl b/src/cowboy_http.erl
index ad2ef75..413c163 100644
--- a/src/cowboy_http.erl
+++ b/src/cowboy_http.erl
@@ -1283,16 +1283,14 @@ stream_terminate(State0=#state{opts=Opts, in_streamid=InStreamID, in_state=InSta
    %% Remove the stream from the state and reset the overriden options.
    {value, #stream{state=StreamState}, Streams}
        = lists:keytake(StreamID, #stream.id, Streams1),
-   State2 = State1#state{streams=Streams, overriden_opts=#{}, flow=infinity},
+   State2 = State1#state{streams=Streams, overriden_opts=#{}},
    %% Stop the stream.
    stream_call_terminate(StreamID, Reason, StreamState, State2),
    Children = cowboy_children:shutdown(Children0, StreamID),
-   %% We reset the timeout if there are no active streams anymore.
-   State = set_timeout(State2#state{streams=Streams, children=Children}, request_timeout),
+   State = State2#state{streams=Streams, children=Children},
    %% We want to drop the connection if the body was not read fully
    %% and we don't know its length or more remains to be read than
    %% configuration allows.
-   %% @todo Only do this if Current =:= StreamID.
    MaxSkipBodyLength = maps:get(max_skip_body_length, Opts, 1000000),
    case InState of
        #ps_body{length=undefined}
@@ -1301,17 +1299,21 @@ stream_terminate(State0=#state{opts=Opts, in_streamid=InStreamID, in_state=InSta
        #ps_body{length=Len, received=Received}
                when InStreamID =:= OutStreamID, Received + MaxSkipBodyLength < Len ->
            terminate(State, skip_body_too_large);
+       #ps_body{} when InStreamID =:= OutStreamID ->
+           stream_next(State#state{flow=infinity});
        _ ->
-           %% Move on to the next stream.
-           NextOutStreamID = OutStreamID + 1,
-           case lists:keyfind(NextOutStreamID, #stream.id, Streams) of
-               false ->
-                   State#state{out_streamid=NextOutStreamID, out_state=wait};
-               #stream{queue=Commands} ->
-                   %% @todo Remove queue from the stream.
-                   commands(State#state{out_streamid=NextOutStreamID, out_state=wait},
-                       NextOutStreamID, Commands)
-           end
+           stream_next(State)
+   end.
+
+stream_next(State=#state{out_streamid=OutStreamID, streams=Streams}) ->
+   NextOutStreamID = OutStreamID + 1,
+   case lists:keyfind(NextOutStreamID, #stream.id, Streams) of
+       false ->
+           State#state{out_streamid=NextOutStreamID, out_state=wait};
+       #stream{queue=Commands} ->
+           %% @todo Remove queue from the stream.
+           commands(State#state{out_streamid=NextOutStreamID, out_state=wait},
+               NextOutStreamID, Commands)
    end.
 
 stream_call_terminate(StreamID, Reason, StreamState, #state{opts=Opts}) ->
```

It should fix both the flow problem from this ticket and a timeout issue I noticed at the same time where the timeout was reset when it shouldn't have been (and could have prevented skipping the body in time, or increased an existing request timeout).

**Comment by huss-kivra** [comment]
We'll apply the patch today. Since we only get the error once a week or so, we'll have to let it run a few weeks before we can be somewhat sure that the issue is fixed.

**Comment by essen** [comment]
No problem. I've pushed a commit with a small cleanup on top of the above patch since I know from tests that it solves at least part of this problem, if not all of it.

**Comment by alexandre-kivra** [comment]
10 days have passed and no error. I think we can assumed the issue is solved. Thank you! Will you create a new tag?

**Comment by essen** [comment]
The plan is to release Cowboy 2.8 sometimes in February.

**Comment by alexandre-kivra** [comment]
@essen the crash on the `update_flow` still appears to have been fixed. No crashes there since we deployed the latest cowboy master.

But now we noticed that we have also been having the `badarith` on `cowboy_http:commands/3`

Example of the crash:

```
Process terminated with reason: {{error,badarith},
 [{erlang,'+',[infinity,8000000],[]},
  {cowboy_http,commands,3,[{file,"/build/example/_build/default/lib/cowboy/src/cowboy_http.erl"},{line,979}]},
  {cowboy_http,loop,1,[{file,"/build/example/_build/default/lib/cowboy/src/cowboy_http.erl"},{line,257}]},
  {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
```

Do you need some extra information to debug, or is there any patch already that you want us to try?

**Comment by essen** [comment]
Did you try with this patch? https://github.com/ninenines/cowboy/issues/1411#issuecomment-549314493

Anyway both have been merged so maybe just use master.

**Comment by alexandre-kivra** [comment]
We are using the latest master (752297b1539b4f74c9329c873f78485a52b5b8bd).

**Comment by essen** [comment]
I guess it hasn't been merged or pushed or something, I don't see it. So apply the patch linked on top and you should not get the error.

**Comment by alexandre-kivra** [comment]
The code changed in the meantime but I imagine this would be the patch with the current master:

```
diff --git a/src/cowboy_http.erl b/src/cowboy_http.erl
index 9f0b865..ccdf615 100644
--- a/src/cowboy_http.erl
+++ b/src/cowboy_http.erl
@@ -970,6 +970,9 @@ commands(State=#state{out_streamid=Current, streams=Streams0}, StreamID, Command
        Streams = lists:keyreplace(StreamID, #stream.id, Streams0,
                Stream#stream{queue=Queue ++ Commands}),
        State#state{streams=Streams};
+%% When we have finished reading the request body, do nothing.
+commands(State=#state{flow=infinity}, StreamID, [{flow, _}|Tail]) ->
+        commands(State, StreamID, Tail);
 %% Read the request body.
 commands(State0=#state{flow=Flow0}, StreamID, [{flow, Size}|Tail]) ->
        %% We must read *at least* Size of data otherwise functions
```

**Comment by essen** [comment]
Yes.

**Comment by essen** [comment]
I've pushed the fix with a test case in 8fc3da2fc31a04281b0ba8dc0cdc1e1650067da5. I think this can be closed now. Thanks everyone for the reports and testing! This will be part of upcoming Cowboy 2.8.
